### ISSUE TITLE
Dialog box text formatted incorrectly?

### ISSUE THREAD

**Issue by CorneliusCornbread** [native-dialog-rs/native-dialog-rs#41]
So I tried this out on KDE on Wayland and doing a basic message dialog no matter what I put in for the message it's malformed:
![image](https://user-images.githubusercontent.com/8294697/235416551-8b6ecfb0-7f15-44ad-be2a-590412451655.png)

message code
```
MessageDialog::new()
      .set_type(MessageType::Error)
      .set_title("Fatal Error")
      .set_text(msg)
      .show_alert()
      .unwrap();
```

**Comment by edfloreshz** [comment]
I'm having this error too.

**Comment by edfloreshz** [comment]
I think the issue might be here:

https://github.com/balthild/native-dialog-rs/blob/d2ddd443f8c01e92dc22cc8132159c1b9598eaca/src/dialog_impl/gnu/message.rs#L56

```rust
/// See https://github.com/qt/qtbase/blob/2e2f1e2/src/gui/text/qtextdocument.cpp#L166
fn convert_qt_text_document(text: &str) -> String {
    text.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\n', "<br>")
        .replace(' ', "&nbsp")
        .replace('\t', "&nbsp;")
}
```

Perhaps removing that method would do the trick? It doesn't seem to be necessary in Rust.

**Comment by CorneliusCornbread** [comment]
> I think the issue might be here:
> 
> https://github.com/balthild/native-dialog-rs/blob/d2ddd443f8c01e92dc22cc8132159c1b9598eaca/src/dialog_impl/gnu/message.rs#L56
> 
> ```rust
> /// See https://github.com/qt/qtbase/blob/2e2f1e2/src/gui/text/qtextdocument.cpp#L166
> fn convert_qt_text_document(text: &str) -> String {
>     text.replace('&', "&amp;")
>         .replace('<', "&lt;")
>         .replace('>', "&gt;")
>         .replace('"', "&quot;")
>         .replace('\n', "<br>")
>         .replace(' ', "&nbsp")
>         .replace('\t', "&nbsp;")
> }
> ```
> 
> Perhaps removing that method would do the trick? It doesn't seem to be necessary in Rust.

Another person mentioned in a pull request that it seems kdialog does not use this formatting anymore. Upon removing this function call yes I can confirm it solved the issue for me.

**Comment by edfloreshz** [comment]
If you're referring to #42, I made that PR 😆

**Comment by balthild** [comment]
Hello! I'm trying to find where and when kdialog changed its behavior, but I find strangely that the behavior should have not been changed. The text preprocessing function in `kdialog` does almost nothing (https://github.com/KDE/kdialog/blame/master/src/utils.cpp#L46), the message box is `kdialog` is just a `KMessageBox` under the hood (https://github.com/KDE/kdialog/blame/master/src/kdialog.cpp#L515), and `KMessageBox` uses `QLabel` to display the text (https://github.com/KDE/kwidgetsaddons/blame/master/src/kmessagebox.cpp#L219) - all these code remains unchanged in recent years, as `git blame` shows. It is really strange that the behavior is not consistent. I'm still trying to figure it out...

**Comment by CorneliusCornbread** [comment]
> Hello! I'm trying to confirm where and when kdialog changed its behavior so that I could escape the text or not for different versions of `kdialog`. However, I find strangely that the behavior should have not been changed.
> 
> The text preprocessing function in `kdialog` does almost nothing (https://github.com/KDE/kdialog/blame/master/src/utils.cpp#L46), the message box is `kdialog` is just a `KMessageBox` under the hood (https://github.com/KDE/kdialog/blame/master/src/kdialog.cpp#L515), and `KMessageBox` uses `QLabel` to display the text (https://github.com/KDE/kwidgetsaddons/blame/master/src/kmessagebox.cpp#L219) - all these code remains unchanged in recent years, as `git blame` shows. It is really strange that the behavior is not consistent.
> 
> I'm still trying to figure it out...

Do you remember roughly which version or when kdialog _did_ have that behavior?

**Comment by balthild** [comment]
> Do you remember roughly which version or when kdialog did have that behavior?

Yes. In fact, I have a Ubuntu 20.04 VM now (thankfully I didn't delete it), and the kdialog in it is `19.12.3`.

![image](https://github.com/native-dialog-rs/native-dialog-rs/assets/2662758/1255342c-f4da-4253-9eba-6e0c7321045b)

**Comment by CorneliusCornbread** [comment]
> > Do you remember roughly which version or when kdialog did have that behavior?
> 
> Yes. In fact, I have a Ubuntu 20.04 VM now (thankfully I didn't delete it), and the kdialog in it is `19.12.3`.
> 
> ![image](https://user-images.githubusercontent.com/2662758/241346136-1255342c-f4da-4253-9eba-6e0c7321045b.png)

What version of Qt is it running? and it looks like its KDE no?

**Comment by balthild** [comment]
Yes, it's KDE.

KDE Plasma: 5.18.8
KDE Framework: 5.68.0
Qt: 5.12.8

**Comment by balthild** [comment]
It should be OK to just remove the escape now, and add it back after figuring it out.

Also, I think I'd better to ask some KDE devs...

**Comment by CorneliusCornbread** [comment]
> > Edit: Moreover, do we want to bother supporting older versions or tell users that they'll have to use say version 22.x.x (22 is the major version I'm running). Depending on how old it is it might not be worth the time investment to figure out which version this was changed in.
> 
> It should be OK to just remove the escape now, and add it back after figuring it out.
> 
> Also, I think I'd better ask some KDE devs...

I agree, however, it should be noted that some formatting does still work. For example bold still works just fine. 
![image](https://github.com/native-dialog-rs/native-dialog-rs/assets/8294697/33a468f8-d128-4586-86fd-e7fd3b8efa40)

**Comment by balthild** [comment]
I also find that `&lt;` and `&gt;` still works. They should be escaped too, as the API of the crate is designed to display the text as-is and don't expect `<>` to become HTML tags.

![image](https://github.com/native-dialog-rs/native-dialog-rs/assets/2662758/9e31d929-217b-43fc-9739-132e2f25309d)

**Comment by CorneliusCornbread** [comment]
> I also find that `&lt;` and `&gt;` still work. They should be escaped too, as the API of the crate is designed to display the text as-is and don't expect `<>` to become HTML tags.
> 
> ![image](https://user-images.githubusercontent.com/2662758/241476127-9e31d929-217b-43fc-9739-132e2f25309d.png)

All this formatting makes me wonder if we should make our own formatting system of some kind and convert it on a per-platform basis.

**Comment by CorneliusCornbread** [comment]
> > Edit: Moreover, do we want to bother supporting older versions or tell users that they'll have to use say version 22.x.x (22 is the major version I'm running). Depending on how old it is it might not be worth the time investment to figure out which version this was changed in.
> 
> It should be OK to just remove the escape now, and add it back after figuring it out.
> 
> Also, I think I'd better ask some KDE devs...

Have you been able to get in contact with KDE devs about which version this might be appearing in?

**Comment by balthild** [comment]
I fired [a post](https://forum.kde.org/viewtopic.php?f=225&t=177742) on the KDE fourm just now. Hope there would be devs see and answer it.

**Comment by CorneliusCornbread** [comment]
> I fired [a post](https://forum.kde.org/viewtopic.php?f=225&t=177742) on the KDE fourm just now. Hope there would be devs see and answer it.

it appears that's the wrong place for the question, as their forums have since moved to https://discuss.kde.org/

**Comment by CorneliusCornbread** [comment]
> I fired [a post](https://forum.kde.org/viewtopic.php?f=225&t=177742) on the KDE fourm just now. Hope there would be devs see and answer it.

have you been able to get an answer yet? I tried to check the discussion myself but it appears I don't have access to it as of late.

**Comment by balthild** [comment]
No. Nobody has answered yet.

https://forum.kde.org/viewtopic.php%3Ff=225&t=177742.html

**Comment by CorneliusCornbread** [comment]
> No. Nobody has answered yet.
> 
> https://forum.kde.org/viewtopic.php%3Ff=225&t=177742.html

it might be better than to ship it with the patch just for versions <=19 and if users who use 20/21 run into issues they raise a bug report and it gets patched accordingly then.

**Comment by balthild** [comment]
Fine. I have published 0.6.4 just now.

**Comment by balthild** [comment]
I posted a question on the new KDE forum, and a KDE dev pointed out that this change might be caused by Qt 5 -> 6 upgrade.

Also, If there's `<html><body></html></body>` sorrunding, the markups take effect again.

https://discuss.kde.org/t/documentation-about-the-xml-entity-escape-used-by-kdialog/6652/2

**Comment by mtkennerly** [comment]
I just ran into this. Adding some info in case it helps:

I reproduced it using kdialog 19.12.3 on Ubuntu 20.04.3 with GNOME 3.36.8. I also got a report about this from someone using kdialog 24.02.1 onOpenSUSE Tumbleweed.

To reproduce, I modified the first `tour` example message to add `\n\nAnother line`, then run `XDG_CURRENT_DESKTOP=KDE cargo run --example tour`.

| version | commit | output |
| - | - | - |
| 0.7.0 | 4b0b159 | ![image](https://github.com/native-dialog-rs/native-dialog-rs/assets/4934192/ef0f5695-b8c6-4a6f-9252-c58ebc8b84ab) |
| 0.6.4 | dcb7930 | ![image](https://github.com/native-dialog-rs/native-dialog-rs/assets/4934192/1ad79ced-1ef0-484a-89ed-52c8455b8bc5) |
| 0.6.3 | c874958 | ![image](https://github.com/native-dialog-rs/native-dialog-rs/assets/4934192/5cd196ed-3b44-448b-b477-1fd26dd07900) |

**Comment by Koranir** [comment]
I'm encountering this bug as well on `KDE Plasma 6.3.4`/`Qt 6.9.0`/`native-dialog 0.8.7`:

![Image](https://github.com/user-attachments/assets/c0850233-2ef2-49e3-9c00-af924df037e1)

**Comment by balthild** [comment]
I changed the condition to escape HTML entities in v0.8.10. Now escaping is skipped for `kdialog` that compiles with Qt6. Could you please check if that works?

**Comment by Koranir** [comment]
![Image](https://github.com/user-attachments/assets/a1de00b2-1de0-4de4-af68-404d198bcf14)

It works with 8.10, thank you.

**Comment by balthild** [comment]
Well it seems that this is still not completely correct... Kdialog 25.04 with Qt6 in openSUSE still escapes string, but kdialog 23.08 with Qt5 in AOSC does not.

I found that it may be determined by the strings' content and not related to versions of kdialog/Qt. For example, `'asdf &apos; asdf<br>asdf'` is treated as HTML but `'<asdf &apos; asdf<br>asdf'` is not. However, strings with '<html>...</html>' are always treated as HTML.

I changed the behavior again, always escaping the strings and adding surrounding '<html>...</html>' in v0.8.11. Could you please test with it again?

**Comment by Koranir** [comment]
Yep, works perfectly with v0.8.11

**Comment by balthild** [comment]
Finally... I think I'm getting close to _the_ solution
